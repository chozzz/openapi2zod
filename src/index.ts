export * from './openapi2zod';