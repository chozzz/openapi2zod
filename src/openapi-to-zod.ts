import { z, ZodSchema, ZodTypeAny } from "zod";
import { OpenAPIV3 } from "openapi-types";
import { openApiSchemaToZod } from "./openapi-schema-to-zod";

/**
 * Parses an OpenAPI schema and returns Zod schemas for all operations.
 * @param openApiSchema - The OpenAPI schema object
 * @returns - A record mapping operation IDs to their Zod schemas
 */
export function parseOpenApiToZod(openApiSchema: OpenAPIV3.Document): Record<string, ZodSchema> {
  const zodSchemas: Record<string, ZodSchema> = {};

  const paths = openApiSchema.paths || {};
  const schemas = openApiSchema.components?.schemas || {};

  // Convert schemas to proper mapping for reference resolution
  const schemaMap: Record<string, OpenAPIV3.SchemaObject> = {};
  for (const [schemaName, schemaDef] of Object.entries(schemas)) {
    if (!('$ref' in schemaDef)) {
      schemaMap[schemaName] = schemaDef;
    }
  }

  for (const [path, pathItem] of Object.entries(paths)) {
    if (!pathItem) continue;

    for (const method of ['get', 'post', 'put', 'patch', 'delete', 'options', 'head', 'trace'] as const) {
      const operation = pathItem[method] as OpenAPIV3.OperationObject | undefined;
      if (!operation) continue;

      const operationId = operation.operationId || `${method.toUpperCase()} ${path}`;

      const parameters = [
        ...(pathItem.parameters || []),
        ...(operation.parameters || []),
      ] as OpenAPIV3.ParameterObject[];

      const requestBody = operation.requestBody;

      // Collect all parameters and request body properties
      const combinedSchemas: Record<string, ZodTypeAny> = {};

      // Process parameters
      for (const param of parameters) {
        const { name, required } = param;
        let paramSchema: ZodTypeAny;

        if ('schema' in param && param.schema) {
          paramSchema = openApiSchemaToZod(param.schema, schemaMap);
        } else {
          console.warn(`Parameter ${name} missing schema.`);
          paramSchema = z.any();
        }
        
        // Attach description to object if available
        if (param?.description && !paramSchema.description) {
          paramSchema = paramSchema.describe(param.description);
        }

        if (!required) {
          paramSchema = paramSchema.optional();
        }

        combinedSchemas[name] = paramSchema;
      }

      // Process requestBody
      let requestBodySchema: ZodTypeAny | undefined = undefined;

      if (requestBody && 'content' in requestBody && requestBody.content) {
        const requestBodyObj = requestBody as OpenAPIV3.RequestBodyObject;
        const jsonContent = requestBodyObj.content["application/json"];
        if (jsonContent && 'schema' in jsonContent && jsonContent.schema) {
          requestBodySchema = openApiSchemaToZod(jsonContent.schema, schemaMap);

          if (requestBodySchema instanceof z.ZodObject) {
            Object.assign(combinedSchemas, requestBodySchema.shape);
          } else {
            combinedSchemas['requestBody'] = requestBodySchema;
          }
        } else {
          console.warn(`Request body for operation ${operationId} missing 'application/json' content schema.`);
        }
      }

      // Create the operation schema
      const operationSchema = z.object(combinedSchemas);

      // Add the schema to the result
      zodSchemas[operationId] = operationSchema;
    }
  }

  return zodSchemas;
}
