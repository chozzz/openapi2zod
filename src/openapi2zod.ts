import { z, ZodSchema, ZodTypeAny } from 'zod'
import { OpenAPIV3 } from 'openapi-types'
import { openApiSchemaToZod } from './openapi-schema-to-zod'
import { flattenZodType } from './utils/flattenZodType'

export function parseOpenApiToZod(openApiSchema: OpenAPIV3.Document): Record<string, ZodSchema> {
  const zodSchemas: Record<string, ZodSchema> = {}

  const paths = openApiSchema.paths || {}
  const schemas = openApiSchema.components?.schemas || {}

  // Convert schemas to proper mapping for reference resolution
  const schemaMap: Record<string, OpenAPIV3.SchemaObject> = {}
  for (const [schemaName, schemaDef] of Object.entries(schemas)) {
    if (!('$ref' in schemaDef)) {
      schemaMap[schemaName] = schemaDef
    }
  }

  for (const [path, pathItem] of Object.entries(paths)) {
    if (!pathItem) continue

    for (const method of [
      'get',
      'post',
      'put',
      'patch',
      'delete',
      'options',
      'head',
      'trace',
    ] as const) {
      const operation = pathItem[method] as OpenAPIV3.OperationObject | undefined
      if (!operation) continue

      const operationId = operation.operationId || `${method.toUpperCase()} ${path}`

      const parameters = [
        ...(pathItem.parameters || []),
        ...(operation.parameters || []),
      ] as OpenAPIV3.ParameterObject[]

      const requestBody = operation.requestBody;

      // Collect all parameters and request body properties
      const combinedSchemas: Record<string, ZodTypeAny> = {}

      // Process parameters
      for (const param of parameters) {
        const { name, required } = param
        let paramSchema: ZodTypeAny

        if ('schema' in param && param.schema) {
          paramSchema = openApiSchemaToZod(param.schema, schemaMap)
        } else {
          console.warn(`Parameter ${name} missing schema.`)
          paramSchema = z.any()
        }

        // Attach description to object if available
        if (param?.description && !paramSchema.description) {
          paramSchema = paramSchema.describe(param.description)
        }

        if (!required) {
          paramSchema = paramSchema.optional()
        }

        combinedSchemas[name] = paramSchema
      }

      // Create parameters schema
      const parametersSchema = z.object(combinedSchemas)

      // Process requestBody
      let requestBodySchema: ZodTypeAny | undefined = undefined

      if (requestBody && 'content' in requestBody && requestBody.content) {
        const requestBodyObj = requestBody as OpenAPIV3.RequestBodyObject
        const jsonContent = requestBodyObj.content['application/json']
        if (jsonContent && 'schema' in jsonContent && jsonContent.schema) {
          requestBodySchema = openApiSchemaToZod(jsonContent.schema, schemaMap)
        } else {
          console.warn(
            `Request body for operation ${operationId} missing 'application/json' content schema.`,
          )
        }
      }

      let operationSchema: ZodTypeAny;

      if (requestBodySchema) {
        const isParametersSchemaEmpty =
          parametersSchema instanceof z.ZodObject &&
          Object.keys(parametersSchema.shape).length === 0

        if (isParametersSchemaEmpty) {
          // No parameters, use requestBodySchema directly
          operationSchema = requestBodySchema
        } else {
          // Merge parametersSchema and requestBodySchema appropriately
          // Existing logic for merging schemas
          const flattenedRequestBodySchema = flattenZodType(requestBodySchema)

          if (flattenedRequestBodySchema instanceof z.ZodObject) {
            operationSchema = parametersSchema.merge(flattenedRequestBodySchema);
          }
          else if (flattenedRequestBodySchema instanceof z.ZodUnion) {
            // Handle ZodUnion by merging parametersSchema with each option in the union
            const mergedOptions = flattenedRequestBodySchema.options.map((option: z.ZodObject<any> | z.ZodAny) => {
              // Instead of directly merging, combine the fields using intersection or custom logic
              return z.intersection(parametersSchema, option);
            });
          
            if (mergedOptions.length >= 2) {
              operationSchema = z.union(mergedOptions as [ZodTypeAny, ZodTypeAny, ...ZodTypeAny[]]);
            } else if (mergedOptions.length === 1) {
              operationSchema = mergedOptions[0];
            } else {
              operationSchema = parametersSchema;
            }
          }
          
          else {
            // For other types of requestBodySchema, use z.intersection
            operationSchema = z.intersection(parametersSchema, flattenedRequestBodySchema);
          }

          // Apply transformation to ensure merged fields are preserved
          operationSchema = operationSchema.transform((input) => {
            // Merge fields from both schemas
            const paramResult = parametersSchema.safeParse(input).success ? input : {};
            const bodyResult = flattenedRequestBodySchema.safeParse(input).success ? input : {};
            
            return { ...paramResult, ...bodyResult };
          });
        }
      } else {
        operationSchema = parametersSchema
      }

      // Attach description or summary to operationSchema if available
      operationSchema = operationSchema.describe(JSON.stringify({
        description: operation.description || operation.summary || '',
        operationId,
        path,
        method,
      }));

      // Add the schema to the result
      zodSchemas[operationId] = operationSchema
    }
  }

  return zodSchemas
}
