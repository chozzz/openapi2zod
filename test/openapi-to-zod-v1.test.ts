import { OpenAPIV3 } from 'openapi-types'
import { z } from 'zod'
import { parseOpenApiToZod } from '../src/openapi-to-zod' // Adjust the import path accordingly
import { zodToJsonSchema } from 'zod-to-json-schema'
import * as openapiJson from './__mocks__/openapi_3_1_0_v1.json'

describe('parseOpenApiToZod v1', () => {
  let zodSchemas: Record<string, z.ZodSchema>

  beforeAll(() => {
    zodSchemas = parseOpenApiToZod(openapiJson as OpenAPIV3.Document)
  })

  test('should return an object with schemas', () => {
    expect(typeof zodSchemas).toBe('object')
    expect(Object.keys(zodSchemas).length).toBeGreaterThan(0)
  })

  test('should generate valid Zod schemas', () => {
    for (const schema of Object.values(zodSchemas)) {
      expect(schema).toBeInstanceOf(z.ZodType)
    }
  })

  test('check zod schema contains description in its definition', () => {
    for (const [operationId, schema] of Object.entries(zodSchemas)) {
      expect(schema._def.description).toBeDefined()
    }
  })

  // Test specific operation: addJiraIssueComment
  test('should generate correct schema for operationId "POST addJiraIssueComment"', () => {
    const schema = zodSchemas['addJiraIssueComment']
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodType)

    // Example valid data
    const validData = {
      issueIdOrKey: 'JIRA-123',
      commentBody: 'This is a test comment.',
    }

    // Example invalid data (missing required field)
    const invalidData = {
      issueIdOrKey: 'JIRA-123',
    }

    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidData)).toThrow()

    // Generate JSON schema from Zod schema
    const jsonSchema = zodToJsonSchema(schema)

    // Use snapshot testing
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "POST addProductiveTimeEntry"', () => {
    const schema = zodSchemas['addProductiveTimeEntry'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      note: 'Worked on project X',
      dateString: 'today',
      hours: 8,
      service_id: '3173137',
      person_id: '404947',
    }

    // Invalid data (missing required fields)
    const invalidDataMissingFields = {
      note: 'Worked on project X',
    }

    // Invalid data (wrong type)
    const invalidDataWrongType = {
      note: 'Worked on project X',
      dateString: 'today',
      hours: 'eight', // Should be integer
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingFields)).toThrow()
    expect(() => schema.parse(invalidDataWrongType)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'addProductiveTimeEntry')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "POST assignJiraIssue"', () => {
    const schema = zodSchemas['assignJiraIssue'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data with assignee
    const validDataWithAssignee = {
      issueIdOrKey: 'JIRA-123',
      assignee: 'user123',
    }

    // Valid data without assignee (should return available assignees)
    const validDataWithoutAssignee = {
      issueIdOrKey: 'JIRA-123',
    }

    // Invalid data (missing issueIdOrKey)
    const invalidData = {
      assignee: 'user123',
    }

    // Validate parsing
    expect(() => schema.parse(validDataWithAssignee)).not.toThrow()
    expect(() => schema.parse(validDataWithoutAssignee)).not.toThrow()
    expect(() => schema.parse(invalidData)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'assignJiraIssue')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "POST commentOnGithubIssue"', () => {
    const schema = zodSchemas['commentOnGithubIssue'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      repoOwner: 'octocat',
      repoName: 'Hello-World',
      issueNumber: 42,
      commentBody: 'This is a test comment.',
    }

    // Invalid data (missing required fields)
    const invalidDataMissingFields = {
      repoOwner: 'octocat',
      issueNumber: 42,
      commentBody: 'This is a test comment.',
    }

    // Invalid data (wrong type)
    const invalidDataWrongType = {
      repoOwner: 'octocat',
      repoName: 'Hello-World',
      issueNumber: 'forty-two', // Should be integer
      commentBody: 'This is a test comment.',
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingFields)).toThrow()
    expect(() => schema.parse(invalidDataWrongType)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'commentOnGithubIssue')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "GET createGmailLabelsAndFilters"', () => {
    const schema = zodSchemas['createGmailLabelsAndFilters'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data with both parameters
    const validDataBothParams = {
      labelsToCreate: 'Important,Work',
      filtersToCreate: 'from:boss@example.com',
    }

    // Valid data with one parameter
    const validDataOneParam = {
      labelsToCreate: 'Personal',
    }

    // Valid data with no parameters (both optional)
    const validDataNoParams = {}

    // Invalid data (wrong type)
    const invalidDataWrongType = {
      labelsToCreate: 123, // Should be string
    }

    // Validate parsing
    expect(() => schema.parse(validDataBothParams)).not.toThrow()
    expect(() => schema.parse(validDataOneParam)).not.toThrow()
    expect(() => schema.parse(validDataNoParams)).not.toThrow()
    expect(() => schema.parse(invalidDataWrongType)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'createGmailLabelsAndFilters')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "POST createGoogleDriveFile"', () => {
    const schema = zodSchemas['createGoogleDriveFile'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      fileName: 'document.txt',
      fileContent: 'This is the content of the file.',
    }

    // Invalid data: missing fileContent
    const invalidDataMissingFileContent = {
      fileName: 'document.txt',
    }

    // Invalid data: fileName is not a string
    const invalidDataFileNameNotString = {
      fileName: 123, // Should be string
      fileContent: 'This is the content of the file.',
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingFileContent)).toThrow()
    expect(() => schema.parse(invalidDataFileNameNotString)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'createGoogleDriveFile')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "DELETE deleteJiraFixVersion"', () => {
    const schema = zodSchemas['deleteJiraFixVersion'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      versionId: '12345',
    }

    // Invalid data: missing versionId
    const invalidDataMissingVersionId = {}

    // Invalid data: versionId is not a string
    const invalidDataVersionIdNotString = {
      versionId: 12345, // Should be a string
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingVersionId)).toThrow()
    expect(() => schema.parse(invalidDataVersionIdNotString)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'deleteJiraFixVersion')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "DELETE deleteJiraIssueComment"', () => {
    const schema = zodSchemas['deleteJiraIssueComment'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      issueIdOrKey: 'JIRA-123',
      commentId: '45678',
    }

    // Invalid data: missing commentId
    const invalidDataMissingCommentId = {
      issueIdOrKey: 'JIRA-123',
    }

    // Invalid data: missing issueIdOrKey
    const invalidDataMissingIssueIdOrKey = {
      commentId: '45678',
    }

    // Invalid data: commentId is not a string
    const invalidDataCommentIdNotString = {
      issueIdOrKey: 'JIRA-123',
      commentId: 45678, // Should be a string
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingCommentId)).toThrow()
    expect(() => schema.parse(invalidDataMissingIssueIdOrKey)).toThrow()
    expect(() => schema.parse(invalidDataCommentIdNotString)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'deleteJiraIssueComment')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "DELETE deleteJiraSprint"', () => {
    const schema = zodSchemas['deleteJiraSprint'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      sprintId: 7890,
    }

    // Invalid data: missing sprintId
    const invalidDataMissingSprintId = {}

    // Invalid data: sprintId is not an integer
    const invalidDataSprintIdNotInteger = {
      sprintId: '7890', // Should be an integer
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingSprintId)).toThrow()
    expect(() => schema.parse(invalidDataSprintIdNotInteger)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'deleteJiraSprint')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "PATCH updateJiraIssue"', () => {
    const schema = zodSchemas['updateJiraIssue'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      issueIdOrKey: 'JIRA-123',
      bodyData: '{"fields": {"summary": "Updated summary"}}',
    }

    // Invalid data: missing issueIdOrKey
    const invalidDataMissingIssueIdOrKey = {
      bodyData: '{"fields": {"summary": "Updated summary"}}',
    }

    // Invalid data: missing bodyData
    const invalidDataMissingBodyData = {
      issueIdOrKey: 'JIRA-123',
    }

    // Invalid data: issueIdOrKey is not a string
    const invalidDataIssueIdOrKeyNotString = {
      issueIdOrKey: 12345, // Should be a string
      bodyData: '{"fields": {"summary": "Updated summary"}}',
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingIssueIdOrKey)).toThrow()
    expect(() => schema.parse(invalidDataMissingBodyData)).toThrow()
    expect(() => schema.parse(invalidDataIssueIdOrKeyNotString)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'updateJiraIssue')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "PATCH updateJiraIssueComment"', () => {
    const schema = zodSchemas['updateJiraIssueComment'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data
    const validData = {
      issueIdOrKey: 'JIRA-123',
      commentId: '45678',
      commentBody: 'Updated comment content.',
    }

    // Invalid data: missing issueIdOrKey
    const invalidDataMissingIssueIdOrKey = {
      commentId: '45678',
      commentBody: 'Updated comment content.',
    }

    // Invalid data: missing commentId
    const invalidDataMissingCommentId = {
      issueIdOrKey: 'JIRA-123',
      commentBody: 'Updated comment content.',
    }

    // Invalid data: missing commentBody
    const invalidDataMissingCommentBody = {
      issueIdOrKey: 'JIRA-123',
      commentId: '45678',
    }

    // Invalid data: commentId is not a string
    const invalidDataCommentIdNotString = {
      issueIdOrKey: 'JIRA-123',
      commentId: 45678, // Should be a string
      commentBody: 'Updated comment content.',
    }

    // Validate parsing
    expect(() => schema.parse(validData)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingIssueIdOrKey)).toThrow()
    expect(() => schema.parse(invalidDataMissingCommentId)).toThrow()
    expect(() => schema.parse(invalidDataMissingCommentBody)).toThrow()
    expect(() => schema.parse(invalidDataCommentIdNotString)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'updateJiraIssueComment')
    expect(jsonSchema).toMatchSnapshot()
  })

  test('should generate correct schema for operationId "PATCH updateJiraSprint"', () => {
    const schema = zodSchemas['updateJiraSprint'] as z.ZodObject<any>
    expect(schema).toBeDefined()
    expect(schema).toBeInstanceOf(z.ZodObject)

    // Valid data with all fields
    const validDataAllFields = {
      sprintId: 7890,
      name: 'Sprint 5',
      startDate: '2023-10-01T00:00:00Z',
      endDate: '2023-10-15T23:59:59Z',
      goal: 'Complete all user stories assigned.',
      state: 'active',
    }

    // Valid data with only required field
    const validDataRequiredOnly = {
      sprintId: 7890,
    }

    // Invalid data: missing sprintId
    const invalidDataMissingSprintId = {
      name: 'Sprint 5',
    }

    // Invalid data: sprintId is not an integer
    const invalidDataSprintIdNotInteger = {
      sprintId: '7890', // Should be an integer
      name: 'Sprint 5',
    }

    // Invalid data: startDate is not a string
    const invalidDataStartDateNotString = {
      sprintId: 7890,
      startDate: 20231001, // Should be a string
    }

    // Validate parsing
    expect(() => schema.parse(validDataAllFields)).not.toThrow()
    expect(() => schema.parse(validDataRequiredOnly)).not.toThrow()
    expect(() => schema.parse(invalidDataMissingSprintId)).toThrow()
    expect(() => schema.parse(invalidDataSprintIdNotInteger)).toThrow()
    expect(() => schema.parse(invalidDataStartDateNotString)).toThrow()

    // Generate JSON schema and snapshot test
    const jsonSchema = zodToJsonSchema(schema, 'updateJiraSprint')
    expect(jsonSchema).toMatchSnapshot()
  })
})
