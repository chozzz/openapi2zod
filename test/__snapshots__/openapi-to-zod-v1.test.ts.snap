// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "DELETE deleteJiraFixVersion" 1`] = `
{
  "$ref": "#/definitions/deleteJiraFixVersion",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "deleteJiraFixVersion": {
      "additionalProperties": false,
      "description": "Deletes a fix version in Jira by its version ID.",
      "properties": {
        "versionId": {
          "description": "The ID of the fix version to delete.",
          "type": "string",
        },
      },
      "required": [
        "versionId",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "DELETE deleteJiraIssueComment" 1`] = `
{
  "$ref": "#/definitions/deleteJiraIssueComment",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "deleteJiraIssueComment": {
      "additionalProperties": false,
      "description": "Delete a specific comment from a Jira issue.",
      "properties": {
        "commentId": {
          "description": "The ID of the comment to delete.",
          "type": "string",
        },
        "issueIdOrKey": {
          "description": "The ID or key of the Jira issue to delete the comment from.",
          "type": "string",
        },
      },
      "required": [
        "issueIdOrKey",
        "commentId",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "DELETE deleteJiraSprint" 1`] = `
{
  "$ref": "#/definitions/deleteJiraSprint",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "deleteJiraSprint": {
      "additionalProperties": false,
      "description": "Deletes a sprint in Jira by its sprint ID.",
      "properties": {
        "sprintId": {
          "description": "The ID of the sprint to delete.",
          "type": "integer",
        },
      },
      "required": [
        "sprintId",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "GET createGmailLabelsAndFilters" 1`] = `
{
  "$ref": "#/definitions/createGmailLabelsAndFilters",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "createGmailLabelsAndFilters": {
      "additionalProperties": false,
      "description": "Create labels and filters in Gmail",
      "properties": {
        "filtersToCreate": {
          "description": "Filters to create",
          "type": "string",
        },
        "labelsToCreate": {
          "description": "Labels to create",
          "type": "string",
        },
      },
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "PATCH updateJiraIssue" 1`] = `
{
  "$ref": "#/definitions/updateJiraIssue",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "updateJiraIssue": {
      "additionalProperties": false,
      "description": "Update Jira issue details, including custom fields like priority, story points, sprint, etc. Requires \`bodyData\` argument to be provided.",
      "properties": {
        "bodyData": {
          "description": "The fields to update in the Jira issue as a JSON string.",
          "type": "string",
        },
        "issueIdOrKey": {
          "description": "The ID or key of the Jira issue to update.",
          "type": "string",
        },
      },
      "required": [
        "issueIdOrKey",
        "bodyData",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "PATCH updateJiraIssueComment" 1`] = `
{
  "$ref": "#/definitions/updateJiraIssueComment",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "updateJiraIssueComment": {
      "additionalProperties": false,
      "description": "Update an existing comment on a specific Jira issue.",
      "properties": {
        "commentBody": {
          "description": "The updated content of the comment.",
          "type": "string",
        },
        "commentId": {
          "description": "The ID of the comment to update.",
          "type": "string",
        },
        "issueIdOrKey": {
          "description": "The ID or key of the Jira issue to update the comment for.",
          "type": "string",
        },
      },
      "required": [
        "issueIdOrKey",
        "commentId",
        "commentBody",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "PATCH updateJiraSprint" 1`] = `
{
  "$ref": "#/definitions/updateJiraSprint",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "updateJiraSprint": {
      "additionalProperties": false,
      "description": "Updates a sprint in Jira with specified parameters such as sprint ID, name, start date, end date, goal, and state. If sprint ID is not provided, returns a list of available sprints.",
      "properties": {
        "endDate": {
          "description": "The new end date of the sprint in ISO8601 format.",
          "type": "string",
        },
        "goal": {
          "description": "The new goal of the sprint.",
          "type": "string",
        },
        "name": {
          "description": "The new name of the sprint.",
          "type": "string",
        },
        "sprintId": {
          "description": "The ID of the sprint to update.",
          "type": "integer",
        },
        "startDate": {
          "description": "The new start date of the sprint in ISO8601 format.",
          "type": "string",
        },
        "state": {
          "description": "The new state of the sprint (e.g., active, closed).",
          "type": "string",
        },
      },
      "required": [
        "sprintId",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "POST addJiraIssueComment" 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "Add a comment to a specific Jira issue.",
  "properties": {
    "commentBody": {
      "description": "The content of the comment.",
      "type": "string",
    },
    "issueIdOrKey": {
      "description": "The ID or key of the Jira issue to add a comment to.",
      "type": "string",
    },
  },
  "required": [
    "issueIdOrKey",
    "commentBody",
  ],
  "type": "object",
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "POST addProductiveTimeEntry" 1`] = `
{
  "$ref": "#/definitions/addProductiveTimeEntry",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "addProductiveTimeEntry": {
      "additionalProperties": false,
      "description": "Updates an existing time entry in Productive by sending a POST request to the Productive API v2.",
      "properties": {
        "dateString": {
          "description": "Date of the time entry, in natural language (e.g., 'tomorrow', 'next Friday'). Chrono-node parses it. Default: 'today at 23:59'",
          "type": "string",
        },
        "hours": {
          "description": "Number of hours logged",
          "type": "integer",
        },
        "note": {
          "description": "Note of the work done for this time entry",
          "type": "string",
        },
        "person_id": {
          "description": " ID of the user associated with the time entry. Leaving it empty will return all available person. Defaults to my person ID.",
          "type": "string",
        },
        "service_id": {
          "description": " ID of the service associated with the time entry. Leaving it empty will return all available services. Defaults to my service ID.",
          "type": "string",
        },
      },
      "required": [
        "note",
        "dateString",
        "hours",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "POST assignJiraIssue" 1`] = `
{
  "$ref": "#/definitions/assignJiraIssue",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "assignJiraIssue": {
      "additionalProperties": false,
      "description": "Assigns a Jira issue to a specified assignee. If the assignee is not provided, it returns the available assignees for the issue.",
      "properties": {
        "assignee": {
          "description": "The account ID or username of the assignee. If not provided, available assignees will be returned.",
          "type": "string",
        },
        "issueIdOrKey": {
          "description": "The ID or key of the Jira issue.",
          "type": "string",
        },
      },
      "required": [
        "issueIdOrKey",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "POST commentOnGithubIssue" 1`] = `
{
  "$ref": "#/definitions/commentOnGithubIssue",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "commentOnGithubIssue": {
      "additionalProperties": false,
      "description": "Add a comment to an existing issue in a specified GitHub repository.",
      "properties": {
        "commentBody": {
          "description": "The content of the comment.",
          "type": "string",
        },
        "issueNumber": {
          "description": "The number of the issue to comment on.",
          "type": "integer",
        },
        "repoName": {
          "description": "The name of the repository.",
          "type": "string",
        },
        "repoOwner": {
          "description": "The owner of the repository.",
          "type": "string",
        },
      },
      "required": [
        "repoOwner",
        "repoName",
        "issueNumber",
        "commentBody",
      ],
      "type": "object",
    },
  },
}
`;

exports[`parseOpenApiToZod v1 should generate correct schema for operationId "POST createGoogleDriveFile" 1`] = `
{
  "$ref": "#/definitions/createGoogleDriveFile",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "createGoogleDriveFile": {
      "additionalProperties": false,
      "description": "Creates and uploads a file to Google Drive with a specified filename and content using Google OAuth token.",
      "properties": {
        "fileContent": {
          "description": "The content to be written into the file.",
          "type": "string",
        },
        "fileName": {
          "description": "The name of the file to be created.",
          "type": "string",
        },
      },
      "required": [
        "fileName",
        "fileContent",
      ],
      "type": "object",
    },
  },
}
`;
